scripts/git-pr

```diff
@@ -1,38 +1,227 @@
 #!/usr/bin/env python
-"""git-pr: fetch a pull request
 
-git pr NR
-git pr NR0 NR1 NR2
-git pr NR0 NR1 NR2 -args -to -git-fetch
+"""Checkout GitHub pull requests locally.
 
-Adds a new branch "pr/NR"
+A small convenience script to checkout pull requests as a local branch
+and clean up once done.
 
+git pr NR
+git pr --push NR
+git pr --done NR
 """
 
-from __future__ import print_function
-import subprocess
 import sys
+import re
+import argparse
+import subprocess
+import urllib.request
+import json
+import shlex
+import traceback
+from contextlib import contextmanager
+
+
+BASE_URL = "git@github.com:"
+
+GITHUB_REST_URL = "https://api.github.com/repos/"
+
+BLACKLIST_BRANCH_NAMES = ("main", "master")
+BLACKLIST_REMOTE_NAMES = ("origin", "upstream", "upstream-writeable")
+
+
+def red(text: str) -> str:
+    """Wrap `text` with bold red ANSII escape code."""
+    return f"\033[31;1m{text}\033[0m"
+
+
+def blue(text) -> str:
+    """Wrap `text` with bold red ANSII escape code."""
+    return f"\033[34m{text}\033[0m"
+
+
+def bold(text) -> str:
+    """Wrap `text` with bold ANSII escape code."""
+    return f"\033[1m{text}\033[0m"
+
+
+def run(cmd: str, *args, check=True, show=False) -> str:
+    """Run a command while handling printing.
+
+    `cmd` should only contain trusted input, while `args` might contain input
+    from untrusted sources, e.g. fetched data from the internet.
+    """
+    if show is True:
+        joined_args = " ".join(shlex.quote(a) for a in args)
+        print(bold(f"$ {cmd} {joined_args}"))
+
+    cmd_parts = shlex.split(cmd)
+    if not cmd_parts or not cmd_parts[0]:
+        raise ValueError(f"cmd appears to be empty: {cmd_parts!r}")
+
+    result = subprocess.run(
+        cmd_parts + list(args),
+        check=check,
+        text=True,
+        stdout=subprocess.PIPE,
+        # Never use shell=True here to prevent shell injections,
+        # as args might contain input fetched from the web
+        shell=False,
+    )
+
+    output = result.stdout
+    if show is True and output:
+        print(result.stdout.strip())
+    return result.stdout
+
+
+def parse_command_line():
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument(
+        "pr_number",
+        metavar="NUMBER",
+        help="Number of the pull request to check out",
+    )
+    parser.add_argument(
+        "-p",
+        "--push",
+        action="store_true",
+        help="Push changes on PR branch to remote (assume existing PR branch)",
```

Comment:
I added the `--push` option as a helper for users that haven't configured their git to not complain about differing local and remote branch names, which is probably most users.

Couldn't think of a better option and this works for me perfectly so far. :shrug: